### PR TITLE
Fixes for Dalvik and DEX

### PR DESCRIPTION
**Pull request by XVilka**
**DO NOT SQUASH**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/rizinbook) with the relevant information (if needed)

**Detailed description**

Sync to the latest radare2, but not everything is imported - only minor changes.

Notable changes are fixes in Dalvik and DEX parsing, dropping old MODE_DIST from the rz-diff code, various memory leaks.

### CODE REVIEW

**Review by thestr4ng3r — DISMISSED** [review]

File: librz/analysis/dwarf_process.c [hunk 1]
```diff
@@ -869,12 +870,12 @@ static const char *map_dwarf_reg_to_ppc64_reg(ut64 reg_num, VariableLocationKind
 	case 6: return "r6";
 	case 7: return "r7";
 	case 8: return "r8";
-	case 9: return "r8";
-	case 10: return "r9";
-	case 11: return "r10";
-	case 12: return "r11";
-	case 13: return "r12";
-	case 14: return "r13";
+	case 9: return "r9";
```
Comment (wargio) [inline librz/analysis/dwarf_process.c:873]: this stuff needs to be rewritten. i don't see any reason to have something like this

Comment (XVilka) [reply librz/analysis/dwarf_process.c:873]: @wargio good point, added a new commit to simplify this code.

File: librz/analysis/dwarf_process.c [hunk 2]
```diff
@@ -857,42 +858,12 @@ static const char *map_dwarf_reg_to_x86_reg(ut64 reg_num, VariableLocationKind *
...
-	case 25: return "r25";
-	case 26: return "r26";
-	case 27: return "r27";
-	case 28: return "r28";
-	case 29: return "r29";
-	case 30: return "r30";
-	case 31: return "r31";
-	default:
+	} else if (reg_num < 32) {
+		return rz_str_newf("r%"PFMT64d, reg_num);
```
Comment (ret2libc) [inline librz/analysis/dwarf_process.c]: this is wrong. It now allocates a string every time and the function returns `const char *` so I doubt it is freed by the caller.

File: librz/cons/dietline.c [hunk 3]
```diff
@@ -1293,6 +1294,7 @@ static void __vi_mode(void) {
 			break;
 		case 'a':
 			__move_cursor_right();
+			break;
```
Comment (ret2libc) [inline librz/cons/dietline.c]: is this right? I think there should be a fallthrough here as well, similar to `A`. If I get it right, this behaves like `a` in vim, so it's similar to `i` but you move first the cursor to the right.

File: librz/cons/dietline.c [hunk 4]
```diff
@@ -1209,6 +1209,7 @@ static void __vi_mode(void) {
 			break;
 		case 'c':
 			I.vi_mode = INSERT_MODE; // goto insert mode
+			break;
```
Comment (ret2libc) [inline librz/cons/dietline.c]: did not check this... is this really right as well?

File: librz/cons/dietline.c [hunk 5]
```diff
@@ -1888,6 +1890,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 						break;
 					case 0x37: // HOME xrvt-unicode
 						rz_cons_readchar();
+						break;
```
Comment (ret2libc) [inline librz/cons/dietline.c]: same here... mmm need to test all these things.

File: librz/core/cfile.c [hunk 6]
```diff
@@ -909,6 +909,7 @@ RZ_API RzCoreFile *rz_core_file_open(RzCore *r, const char *file, int flags, ut6
 	if (rz_io_is_listener(r->io)) {
 		rz_core_serve(r, fd);
 		rz_io_desc_free(fd);
+		fd = NULL;
```
Comment (ret2libc) [inline librz/core/cfile.c]: why is this necessary? It seems it's a dead assignment, because `fd`is not used in `beach`.

File: librz/core/disasm.c [hunk 7]
```diff
@@ -2416,7 +2416,6 @@ static void ds_update_ref_lines(RDisasmState *ds) {
 		free(ds->prev_line_col);
 		ds->refline = strdup("");
 		ds->refline2 = strdup("");
-		ds->prev_line_col = strdup("");
```
Comment (ret2libc) [inline librz/core/disasm.c:2419]: I'm not sure why this... Do we know?

Comment (XVilka) [reply librz/core/disasm.c:2419]: It is assigned NULL later in the function.

File: librz/core/cfile.c (hunk 6 above)
Comment (XVilka) [reply librz/core/cfile.c]: Good catch! Removed that line.

File: librz/cons/dietline.c (hunk 3 above)
Comment (XVilka) [reply librz/cons/dietline.c]: Yes, you are right. I added a fall-through comment to indicate that.

File: librz/cons/dietline.c (hunk 4 above)
Comment (XVilka) [reply librz/cons/dietline.c]: Yes, was wrong as well. Added a fall-through comment to indicate that.

File: librz/cons/dietline.c (hunk 5 above)
Comment (XVilka) [reply librz/cons/dietline.c]: Added a fall-through comment to indicate that.

File: librz/analysis/dwarf_process.c (hunk 2 above)
Comment (wargio) [reply librz/analysis/dwarf_process.c]: i agree. @XVilka can you open an issue to refactor this? because this needs more work.

File: librz/core/disasm.c (hunk 7 above)
Comment (ret2libc) [reply librz/core/disasm.c:2419]: Right!

File: librz/analysis/dwarf_process.c (hunk 2 above)
Comment (XVilka) [reply librz/analysis/dwarf_process.c]: @ret2libc the simplification will involve changing more than just one function. Are you ok with merging this fix and do the rework of that file later?

File: librz/core/canalysis.c [hunk 8]
```diff
@@ -2385,14 +2385,15 @@ RZ_API void rz_core_analysis_callgraph(RzCore *core, ut64 addr, int fmt) {
 	case RZ_GRAPH_FORMAT_DOT:
 		if (!is_html) {
 			const char *gv_edge = rz_config_get(core->config, "graph.gv.edge");
-			const char *gv_node = rz_config_get(core->config, "graph.gv.node");
+			char *gv_node = strdup(rz_config_get(core->config, "graph.gv.node"));
```
Comment (ret2libc) [inline librz/core/canalysis.c:2388]: this is going to leak because it is never freed (apart when it is reallocated).

**Review by ret2libc — COMMENTED** [review]
LGTM apart from that small thing

**Review by ret2libc — APPROVED** [review]
